content/helpers.go

```diff
@@ -273,3 +330,58 @@ func copyWithBuffer(dst io.Writer, src io.Reader) (written int64, err error) {
 	}
 	return
 }
+
+func newTrylock() *trylock {
+	return &trylock{
+		locks: make(map[string]*lockItem),
+	}
+}
+
+type trylock struct {
+	sync.Mutex
+
+	locks map[string]*lockItem
+}
+
+type lockItem struct {
+	waiter *semaphore.Weighted
+	cnt    int
+}
+
+func (l *trylock) lock(ctx context.Context, key string) error {
+	l.Lock()
+	w, ok := l.locks[key]
+	if !ok {
+		l.locks[key] = &lockItem{
+			waiter: semaphore.NewWeighted(1),
+		}
+		w = l.locks[key]
+	}
+	w.cnt++
+	l.Unlock()
+
+	if err := w.waiter.Acquire(ctx, 1); err != nil {
+		l.Lock()
+		w.cnt--
+		if w.cnt <= 0 {
+			delete(l.locks, key)
+		}
```

Comment:
Would it be negative?

Reply:
Basically, it is impossible. Just check `<=0` in case. :p

Reply:
How about checking zero in this `if` and returning an error on the negative case?